benches/nixpkgs/lists.ncl

```diff
@@ -116,4 +116,554 @@
+    => let found = array.filter pred list in
+       let len = array.length found in
+       if len == 0 then def
+       else if len != 1 then multiple
+       else array.head found,
+
+  findFirst: forall a. (a -> Bool) -> a -> Array a -> a
+  | doc m%"
+      Find the first element in the list matching the specified
+     predicate or return `default` if no such element exists.
+
+     Example:
+       findFirst (fun x => x > 3) 7 [ 1, 6, 4 ]
+       >> 6
+       findFirst (fun x => x > 9) 7 [ 1, 6, 4 ]
+       >> 7
+  "%m
+  = fun
+    # Predicate
+    pred
+    # Default value to return
+    def
+    # Input list
+    list
+    => let found = array.filter pred list in
+       if found == [] then def else array.head found,
+
+  any: forall a. (a -> Bool) -> Array a -> Bool
+  | doc m%"
+      Return true if function `pred` returns true for at least one
+     element of `list`.
+
+     Example:
+       any builtin.is_str [ 1, "a", { } ]
+       >> true
+       any builtin.is_str [ 1, { } ]
+       >> false
+  "%m
+  = fun pred => foldr (fun x y => if pred x then true else y) false,
+
+  all: forall a. (a -> Bool) -> Array a -> Bool
+  | doc m%"
+      Return true if function `pred` returns true for all elements of
+     `list`.
+
+     Example:
+       all (fun x => x < 3) [ 1, 2 ]
+       >> true
+       all (fun x => x < 3) [ 1, 2, 3 ]
+       >> false
+  "%m
+  = fun pred => foldr (fun x y => if pred x then y else false) true,
+
+  count: (a -> Bool) -> Array a -> Num
+  | doc m%"
+      Count how many elements of `list` match the supplied predicate
+     function.
+
+     Example:
+       count (fun x => x == 3) [ 3, 2, 3, 4, 6 ]
+       >> 2
+  "%m
+  = fun
+    # Predicate
+    pred => foldl_ (fun c x => if pred x then c + 1 else c) 0,
+
+  optional: forall a. Bool -> a -> Array a
+  | doc m%"
+      Return a singleton list or an empty list, depending on a boolean
+     value.  Useful when building lists with optional elements
+     (e.g. `++ optional (system == "i686-linux") firefox').
+
+     Example:
+       optional true "foo"
+       >> [ "foo" ]
+       optional false "foo"
+       >> [ ]
+  "%m
+  = fun cond elem => if cond then [elem] else [],
+
+  optionals: forall a. Bool -> Array a -> Array a
+  | doc m%"
+      Return a list or an empty list, depending on a boolean value.
+
+     Example:
+       optionals true [ 2, 3 ]
+       >> [ 2, 3 ]
+       optionals false [ 2, 3 ]
+       >> [ ]
+  "%m
+  = fun
+    # Condition
+    cond
+    # List to return if condition is true
+    elems
+    => if cond then elems else [],
+
+
+  toList: Dyn -> Array Dyn
+  | doc m%"
+      If argument is a list, return it, else, wrap it in a singleton
+     list.  If you're using this, you should almost certainly
+     reconsider if there isn't a more "well-typed" approach.
+
+     Example:
+       toList [ 1, 2 ]
+       >> [ 1, 2 ]
+       toList "hi"
+       >> [ "hi "]
+  "%m
+  = fun x => if builtin.is_array x then (x | Array Dyn) else [x],
```

Comment:
Ah, interesting example of how flow typing could be useful here (would eliminate the need for a contract application).